apps/documents/serializers.py

```diff
@@ -6,55 +6,105 @@
+            Paragraph.objects.create(chapter=chapter,
+                                     **paragraph_data)
 
         return chapter
 
     def update(self, instance, validated_data):
-        instance.name = validated_data['name']
-        instance.save()
-        paragraphs = validated_data.pop('paragraphs')
+        # Update the chapter
+        paragraphs_data = validated_data.pop('paragraphs')
+        chapter = super(ChapterSerializer, self)\
+            .update(instance, validated_data)
 
-        paragraph_ids = [item['id'] for item in paragraphs if 'id' in item]
-        instance.paragraphs.exclude(id__in=paragraph_ids).delete()
+        # Delete removed paragraphs from the database
+        paragraph_ids = [item['id']
+                         for item in paragraphs_data
+                         if 'id' in item]
+        chapter.paragraphs.exclude(id__in=paragraph_ids).delete()
+
+        # Update or create the chapters
+        for weight, paragraph_data in enumerate(paragraphs_data):
+            paragraph_data['weight'] = weight
+            if 'id' in paragraph_data:
+                paragraph = Paragraph.objects.get(id=paragraph_data['id'])
+                assert paragraph.chapter == chapter
+                Paragraph.objects\
+                    .filter(id=paragraph.id)\
+                    .update(**paragraph_data)
 
-        for paragraph in paragraphs:
-            paragraph['chapter'] = instance
-            if 'id' in paragraph:
-                instance.paragraphs.filter(id=paragraph['id'])\
-                                   .update(**paragraph)
             else:
-                instance.paragraphs.create(**paragraph)
+                Paragraph.objects.create(chapter=chapter,
+                                         **paragraph_data)
+
+        return chapter
+
+
+class DocumentSerializer(serializers.Serializer):
+    chapters = ChapterSerializer(many=True)
+
+    def create(self, validated_data):
+        chapters_data = validated_data['chapters']
+        chapters = list(self._create_or_update(chapters_data))
+        return {
+            'chapters': chapters
+        }
+
+    def _create_or_update(self, validated_data):
+        chapter_list_serializer = self.fields['chapters']
+        chapter_serializer = chapter_list_serializer.child
 
-        return instance
+        chapter_ids = [chapter['id']
+                       for chapter in validated_data
+                       if 'id' in chapter]
+        Chapter.objects \
+            .filter(module_id=self.context['module_pk']) \
+            .exclude(id__in=chapter_ids) \
+            .delete()
+
+        for weight, chapter_data in enumerate(validated_data):
+            chapter_data['weight'] = weight
+            if 'id' in chapter_data:
+                instance = Chapter.objects.get(id=chapter_data['id'])
+                assert instance.module_id == self.context['module_pk']
+                yield chapter_serializer.update(instance, chapter_data)
```

Comment:
i was to lazy to build the resulting arrays with "result = [] ... result.append(instance) ... return result"
but i can change it